TulparAudio/internal/include/tulpar/internal/SourceCollection.hpp

```diff
@@ -0,0 +1,333 @@
+/*
+* Copyright (C) 2018 by Godlike
+* This code is licensed under the MIT license (MIT)
+* (http://opensource.org/licenses/MIT)
+*/
+
+#ifndef TULPAR_INTERNAL_SOURCE_COLLECTION_HPP
+#define TULPAR_INTERNAL_SOURCE_COLLECTION_HPP
+
+#include <tulpar/internal/Collection.hpp>
+#include <tulpar/internal/Context.hpp>
+
+#include <tulpar/internal/BufferCollection.hpp>
+
+#include <tulpar/audio/Buffer.hpp>
+#include <tulpar/audio/Source.hpp>
+
+#include <tulpar/TulparAudio.hpp>
+
+#include <array>
+#include <unordered_map>
+#include <vector>
+
+namespace tulpar
+{
+namespace internal
+{
+
+/** @brief  Handles basic source id operations */
+struct OpenAVSourceHandler
+{
+    /** @brief  Generates a collection of source handles via OpenAL
+     *
+     *  @param  batchSize   number of handles to generate
+     *
+     *  @return a collection of newly generated source handles
+     */
+    static Collection<audio::Source>::Handles Generate(uint32_t batchSize);
+
+    /** @brief  Resets given handle
+     *
+     *  Stops given source
+     *
+     *  @param  handle  valid source handle
+     */
+    static void Reclaim(Collection<audio::Source>::Handle handle);
+
+    /** @brief  Returns given collection of handles to OpenAL
+     *
+     *  @param  handle  collection of valid source handles
+     */
+    static void Delete(Collection<audio::Source>::Handles const& handle);
+};
+
+/** @brief  Collection working with audio sources */
+class SourceCollection
```

Comment:
This class has a lot of possible free utility functions for working with sources. Maybe consider leaving it as a specialized container and creating utility namespace for working with sources? I think that it is generally will make this class a lot less clamped with various interfaces that do change its state.

Reply:
Could you give an example what would be left in this class and what would be moved? I can see only `SourceCollection::InheritCollection` as some collection-related method, everything else appears to work with handles.

Would you create several utility classes just to bundle some functions? Could you give an example how you'd like it to be split?